device/ArithmeticOperations.cpp

```diff
@@ -27,52 +40,115 @@ PackedFloat Multiply(PackedFloat const &a, PackedFloat const &b) {
     ap_uint<kMantissaBits> b_mantissa(*reinterpret_cast<ap_uint<kMantissaBits> const *>(b.mantissa));
-    const bool a_is_larger = a.exponent > b.exponent;
-    const bool b_is_larger = !a_is_larger;
-    const bool a_is_zero = a_mantissa == 0;
-    const bool b_is_zero = b_mantissa == 0;
-    const Exponent res_exponent = ((a_is_larger && !a_is_zero) || b_is_zero) ? a.exponent : b.exponent;
-    const Exponent shift_c = (b_is_larger && !b_is_zero) ? (b.exponent - a.exponent) : 0;
-    const Exponent shift_m = (a_is_larger && !a_is_zero) ? (a.exponent - b.exponent) : 0;
-    // Optionally shift by 1, 2, 4, 8, 16... log2(B), such that all bits have eventually traveled to
+
+#ifndef HLSLIB_SYNTHESIS
+    // We better not be getting subnormal inputs
+    assert(a.IsZero() || IsMostSignificantBitSet(a_mantissa));
+    assert(a.IsZero() || IsMostSignificantBitSet(b_mantissa));
+#endif
+
+    const bool exp_are_equal = (a.exponent == b.exponent);
+    const bool a_exp_is_larger = (a.exponent > b.exponent);
+    const bool a_mant_is_zero = a_mantissa == 0;
+    const bool b_mant_is_zero = b_mantissa == 0;
+    const bool a_is_larger = (!a_mant_is_zero) && (a_exp_is_larger || (exp_are_equal && a.mantissa > b.mantissa));
+    const bool subtraction = a.sign != b.sign;
+    Exponent res_exponent = ((a_exp_is_larger && !a_mant_is_zero) || b_mant_is_zero) ? a.exponent : b.exponent;
+    const Exponent shift_c = (!a_exp_is_larger && !b_mant_is_zero) ? (b.exponent - a.exponent) : 0;
+    const Exponent shift_m = (a_exp_is_larger && !a_mant_is_zero) ? (a.exponent - b.exponent) : 0;
+
+#ifndef HLSLIB_SYNTHESIS
+    // Turns out Xilinx allows signed shifts!
+    assert(shift_m >= 0);
+    assert(shift_c >= 0);
+#endif
+
+    // Optionally shift right by 1, 2, 4, 8, 16... log2(B), such that all bits have eventually traveled to
     // their designated position.
-    const int kNumStages = hlslib::ConstLog2(kBits);
-ShiftStages:
-    for (int i = 0; i < kNumStages; ++i) {
-#pragma HLS UNROLL
-        a_mantissa = ((shift_c & (1 << i)) == 0) ? a_mantissa : (a_mantissa >> (1 << i));
-        b_mantissa = ((shift_m & (1 << i)) == 0) ? b_mantissa : (b_mantissa >> (1 << i));
-    }
+    // This is unsigned
+    a_mantissa = a_mantissa >> shift_c;
+    b_mantissa = b_mantissa >> shift_m;
+//     const int kNumStages = hlslib::ConstLog2(kMantissaBits);
+// ShiftStages:
+//     for (int i = 0; i < kNumStages; ++i) {
+// #pragma HLS UNROLL
+//         a_mantissa = ((shift_c & (1 << i)) == 0) ? a_mantissa : (a_mantissa >> (1 << i));
+//         b_mantissa = ((shift_m & (1 << i)) == 0) ? b_mantissa : (b_mantissa >> (1 << i));
+//     }
     // Finally zero out the mantissas if they are shifted by more than the precision
-    a_mantissa = shift_c >= kBits ? decltype(a_mantissa)(0) : a_mantissa;
-    b_mantissa = shift_m >= kBits ? decltype(b_mantissa)(0) : b_mantissa;
+    // Is this necessary?
```

Comment:
It was necessary with the old method of shifting, since the unrolled loop only implements "useful" shifts. I think you can comment it out along with the old shifter, assuming it still passes tests